File: frontend/src/services/caseReview.ts
import { apiClient } from './api'

export interface ReviewCasePayload {
  case_type: 'api' | 'functional'
  case_id: number
}

export interface CreateReviewPayload {
  title: string
  deadline?: string
  reviewer_ids: number[]
  cases: ReviewCasePayload[]
}

export interface UpdateReviewPayload {
  title?: string
  deadline?: string
  status?: string
}

export interface SubmitItemPayload {
  result: 'approved' | 'rejected' | 'needs_change'
  comment?: string
}

export interface ReviewMemberInfo {
  user_id: number
  username: string | null
  status: 'pending' | 'completed'
}

export interface ReviewListItem {
  id: number
  title: string
  status: 'active' | 'completed' | 'expired'
  deadline: string | null
  created_at: string
  created_by: number | null
  creator_name: string | null
  total_cases: number
  total_items: number
  approved_count: number
  pass_rate: number
  completed_members: number
  total_members: number
  members: ReviewMemberInfo[]
}

export interface ReviewCaseView {
  case_type: 'api' | 'functional'
  case_id: number
  case_name: string
  priority: string
  my_item: {
    id: number
    result: string
    comment: string | null
    reviewed_at: string | null
  } | null
  others_opinions: {
    reviewer_id: number
    reviewer_name: string | null
    result: string
    comment: string | null
    reviewed_at: string | null
  }[]
}

export interface ReviewDetail extends ReviewListItem {
  cases: ReviewCaseView[]
}

export const caseReviewApi = {
  list: (projectId: number, status?: string) =>
    apiClient.get<ReviewListItem[]>(`/api/v1/projects/${projectId}/case-reviews`, {
      params: status ? { status } : undefined,
    }),

  create: (projectId: number, payload: CreateReviewPayload) =>
    apiClient.post<ReviewListItem>(`/api/v1/projects/${projectId}/case-reviews`, payload),

  get: (reviewId: number) =>
    apiClient.get<ReviewDetail>(`/api/v1/case-reviews/${reviewId}`),

  update: (reviewId: number, payload: UpdateReviewPayload) =>
    apiClient.put<ReviewListItem>(`/api/v1/case-reviews/${reviewId}`, payload),

  delete: (reviewId: number) =>
    apiClient.delete(`/api/v1/case-reviews/${reviewId}`),

  submitItem: (reviewId: number, itemId: number, payload: SubmitItemPayload) =>
    apiClient.patch(`/api/v1/case-reviews/${reviewId}/items/${itemId}`, payload),

  complete: (reviewId: number) =>
    apiClient.post(`/api/v1/case-reviews/${reviewId}/complete`),
}

File: backend/app/models/case_review.py
from datetime import datetime
from sqlalchemy import Column, Integer, String, Text, Date, DateTime, ForeignKey
from sqlalchemy.orm import relationship
from app.core.database import Base


class CaseReview(Base):
    __tablename__ = "case_reviews"

    id         = Column(Integer, primary_key=True, index=True)
    project_id = Column(Integer, ForeignKey("projects.id", ondelete="CASCADE"), nullable=False)
    title      = Column(String(200), nullable=False)
    status     = Column(String(20), default="active")  # active / completed / expired
    deadline   = Column(Date, nullable=True)
    created_by = Column(Integer, ForeignKey("users.id"), nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)

    members = relationship("CaseReviewMember", back_populates="review", cascade="all, delete-orphan")
    items   = relationship("CaseReviewItem",   back_populates="review", cascade="all, delete-orphan")
    creator = relationship("User", foreign_keys=[created_by])


class CaseReviewMember(Base):
    __tablename__ = "case_review_members"

    id        = Column(Integer, primary_key=True, index=True)
    review_id = Column(Integer, ForeignKey("case_reviews.id", ondelete="CASCADE"), nullable=False)
    user_id   = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    status    = Column(String(20), default="pending")  # pending / completed

    review = relationship("CaseReview", back_populates="members")
    user   = relationship("User", foreign_keys=[user_id])


class CaseReviewItem(Base):
    __tablename__ = "case_review_items"

    id          = Column(Integer, primary_key=True, index=True)
    review_id   = Column(Integer, ForeignKey("case_reviews.id", ondelete="CASCADE"), nullable=False)
    case_type   = Column(String(20), nullable=False)     # 'api' | 'functional'
    case_id     = Column(Integer, nullable=False)
    reviewer_id = Column(Integer, ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    result      = Column(String(20), default="pending")  # pending / approved / rejected / needs_change
    comment     = Column(Text, nullable=True)
    reviewed_at = Column(DateTime, nullable=True)

    review   = relationship("CaseReview", back_populates="items")
    reviewer = relationship("User", foreign_keys=[reviewer_id])
